src/renderer/routes/flow_chart/components/NodeContextMenu.tsx

```diff
@@ -85,26 +88,46 @@ export default function ContextMenu({
   };
 
   const editPythonCode = async () => {
-    console.log(fullPath);
-    await window.api.openEditorWindow(fullPath);
+    try {
+      authenticate(user);
+      console.log(fullPath);
+      await window.api.openEditorWindow(fullPath);
+    } catch (e) {
+      //
+    }
   };
 
   const duplicate = () => {
-    const node = getNode(id);
-    if (!node) {
-      return;
+    try {
+      authenticate(user);
+      const node = getNode(id);
+      if (!node) {
+        return;
+      }
+      duplicateNode(node);
+    } catch (e) {
+      //
     }
-    duplicateNode(node);
   };
 
   const openInVSC = async () => {
-    await window.api.openLink(`vscode://file/${fullPath}`);
+    try {
+      authenticate(user);
+      await window.api.openLink(`vscode://file/${fullPath}`);
+    } catch (e) {
+      //
+    }
   };
 
   const deleteNode = useCallback(() => {
-    setNodes((nodes) => nodes.filter((node) => node.id !== id));
-    setEdges((edges) => edges.filter((edge) => edge.source !== id));
-  }, [id, setNodes, setEdges]);
+    try {
+      authenticate(user);
```

Comment:
Instead of calling this authenticate function every time user makes some changes, just have a provider along with a hook that you can get the state from. Then you just need to update the authentication state only when there is a change.

Reply:
@itsjoeoui  This function is used to determine whether user is permitted to do that action or not. I can't seem a way to handle it way better then this using Provider+hook. You need to either call this function repeatedly or you need to write this code at starting of each function that requires checking user role:
```
if(user.role !== Roles.admin){
   toast.message("Action is not allowed!)
   return
 }
 ```
 I can replace `authenticate(user)` function call with above piece of code if you wish to...
 
 Please let me know if I misunderstood or missing something

Reply:
I see, I will wait to see what @39bytes thinks as well 👍

Reply:
(messaged on Slack about this already)